app/services/sarses_generator.py

```diff
@@ -0,0 +1,135 @@
+# pylint: disable=no-member
+
+"""This module contains logic for composing marketplace DB dump,
+    recommendations and user actions into SARSes
+    needed for Recommender system's replay buffer.
+
+    marketplace DB dump ---\
+                            \
+    recommendations ---------+----> SARSes
+                            /
+    user actions ----------/
+
+    It assumes that above data are stored in database before call.
+    """
+
+import itertools
+from app.db.mongo_models import UserAction, Recommendation, State, Sars
+from app.reward_mapping import ua_to_reward_id
+
+
+RECOMMENDATION_PAGES_IDS = ["catalogue_services_list"]
+
+
+def _tree_collapse(user_action):
+    """Collapse tree of user actions rooted in user_action
+    into list of reward ids"""
+    # Accumulator
+    rewards = [ua_to_reward_id(user_action)]
+
+    # Recursion stop condition
+    if user_action.action.order:
+        return rewards
+
+    if user_action.target.page_id in RECOMMENDATION_PAGES_IDS:
+        return rewards
+
+    target_id = user_action.target.visit_id
+    children = list(UserAction.objects(source__visit_id=target_id))
+    rewards += list(itertools.chain(*[_tree_collapse(child) for child in children]))
+
+    return rewards
+
+
+def _get_clicked_services_and_reward(recommendation, root_uas):
+    """Collapse root user actions trees to:
+    - clicked_services_after
+    - rewards lists (reward)"""
+
+    reward = []
+    clicked_services_after = []
+    for service in recommendation.services:
+        service_root_uas = root_uas(
+            source__visit_id=recommendation.visit_id, source__root__service=service
+        )
+        if service_root_uas:
+            clicked_services_after.append(service)
+        service_rewards = list(
+            itertools.chain(
+                *[
+                    _tree_collapse(service_root_ua)
+                    for service_root_ua in service_root_uas
+                ]
+            )
+        )
+        reward.append(service_rewards)
+
+    return clicked_services_after, reward
+
+
+def _find_root_uas_before(root_uas, recommendation):
+    """Find all user's root actions that user took before
+    current recommendation
+    """
+    if recommendation.logged_user:
+        root_uas_before = root_uas(
+            user=recommendation.user, timestamp__lt=recommendation.timestamp
+        )
+    else:
+        root_uas_before = root_uas(
+            unique_id=recommendation.unique_id,
+            timestamp__lt=recommendation.timestamp,
+        )
+
+    return root_uas_before.order_by("+timestamp")
+
+
+def _ruas2services(root_uas):
+    """This function maps root user actions to services"""
+    return list(map(lambda root_ua: root_ua.source.root.service, root_uas))
```

Comment:
if you're doing `list(map(lambda ...` it's much better and more pythonic to use list comprehension - above line can be rewritten as:
```
return [root_ua.source.root.service for root_ua in root_uas]
```